packages/catalog-process/src/services/catalogService.ts

```diff
@@ -3315,39 +3313,41 @@ async function applyVisibilityToEService(
   authData: UIAuthData | M2MAuthData | M2MAdminAuthData,
   readModelService: ReadModelService
 ): Promise<EService> {
-  if (
-    hasAtLeastOneUserRole(authData, [
-      userRole.ADMIN_ROLE,
-      userRole.API_ROLE,
-      userRole.SUPPORT_ROLE,
-    ]) &&
-    authData.organizationId === eservice.producerId
-  ) {
-    return eservice;
-  }
+  if (hasRoleToAccessInactiveDescriptors(authData)) {
+    /* Inactive descriptors are visible only if both conditions are met:
+       1) The request is made with a role that can access inactive descriptors.
+          (see the condition above)
+       2) The request originates from the producer tenant or a delegate producer tenant.
+          (see the following code)
+    */
+    if (authData.organizationId === eservice.producerId) {
+      return eservice;
+    }
 
-  const producerDelegation = await readModelService.getLatestDelegation({
-    eserviceId: eservice.id,
-    delegateId: authData.organizationId,
-    kind: delegationKind.delegatedProducer,
-  });
+    const producerDelegation = await readModelService.getLatestDelegation({
+      eserviceId: eservice.id,
+      states: [delegationState.active],
+      kind: delegationKind.delegatedProducer,
+    });
 
-  if (producerDelegation?.state === delegationState.active) {
-    return eservice;
+    if (authData.organizationId === producerDelegation?.delegateId) {
+      return eservice;
+    }
   }
 
-  const hasNoActiveDescriptor = eservice.descriptors.every(
-    isNotActiveDescriptor
-  );
-
-  if (!producerDelegation && hasNoActiveDescriptor) {
-    throw eServiceNotFound(eservice.id);
+  /* If the conditions above are not met:
+    - we filter out the draft descriptors.
+    - we throw a not found error if there are no active descriptors
+  */
+  const hasActiveDescriptors = eservice.descriptors.some(isActiveDescriptor);
+  if (hasActiveDescriptors) {
```

Comment:
```suggestion
  const hasNoActiveDescriptors = !eservice.descriptors.some(isActiveDescriptor);
  if (hasNoActiveDescriptors) {
     throw eServiceNotFound(eservice.id);
```
WYT on throwing inside the condition instead of at the end of the method?

I find it clearer to return as the natural end of the function and to throw in a condition instead of the opposite

Reply:
This function has multiple return paths: two above and one here. In cases like this, I prefer to follow this switch-like pattern: in case A --> return A; in case B --> return B; otherwise, throw at the end.

Also, throwing at the end makes it the default behaviour, which I think is safer in cases where we want to limit access to resources: we whitelist all the good cases and throw in otherwise.